### PR TITLE
[linting] Enable eslint/no-unsafe-argument

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Enable "@typescript-eslint/no-unsafe-argument": "error" ([docs](https://typescript-eslint.io/rules/no-unsafe-assignment/)) as a linting rule. 

It flags any assignments from any to a more specific type and forces us to explicitly handle use of `any`. 
Since we have a few places where this flags, I went through those cases and either fixed or explicitly ignored (only in unit tests and very old code where fixing is riskier).

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at b3ebeae</samp>

This pull request improves the reliability, readability, and type safety of the code in the `gitpod-db` and `gitpod-protocol` components, by adding and fixing ESLint rules, error handling, type annotations, and test cases. It also refactors and simplifies some of the database and proxy-factory modules.

</details>

## Related Issue(s)

Fixes #

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-lint-unsafe-arg</li>
	<li><b>🔗 URL</b> - <a href="https://se-lint-unsafe-arg.preview.gitpod-dev.com/workspaces" target="_blank">se-lint-unsafe-arg.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - se-lint-unsafe-arg-gha.18886</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-se-lint-unsafe-arg%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts [hunk 1]
```diff
@@ -167,17 +169,10 @@ export class TypeORMWorkspaceDBImpl extends TransactionalDBImpl<WorkspaceDB> imp
     }
 
     public async findByInstanceId(instanceId: string): Promise<MaybeWorkspace> {
+        const instanceRepo = await this.getWorkspaceInstanceRepo();
+        const instance = await instanceRepo.findOne(instanceId);
         const workspaceRepo = await this.getWorkspaceRepo();
-        const maybeRawWorkspaces = (await workspaceRepo.query(
```
Comment (svenefftinge) [inline components/gitpod-db/src/typeorm/workspace-db-impl.ts:171]: This was exposing broken workspaces. I fixed this and added a test

File: components/gitpod-protocol/src/messaging/proxy-factory.ts [hunk 2]
```diff
@@ -217,11 +223,13 @@ export class JsonRpcProxyFactory<T extends object> implements ProxyHandler<T> {
...
                         try {
                             if (isNotify) {
+                                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                                 connection.sendNotification(p.toString(), ...args);
                                 resolve(undefined);
                             } else {
+                                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                                 const resultPromise = connection.sendRequest(p.toString(), ...args) as Promise<any>;
-                                resultPromise.catch((err: any) => reject(err)).then((result: any) => resolve(result));
+                                resultPromise.then((result: any) => resolve(result)).catch((err: any) => reject(err));
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/messaging/proxy-factory.ts:232]: I believe this was the wrong order. cc @akosyakov @geropl

File: components/server/src/prebuilds/github-app.ts [hunk 3]
```diff
@@ -570,7 +570,7 @@ export class GithubApp {
 
         const newBody = body + `\n\n${button}\n\n`;
         const updatePrPromise = ctx.octokit.pulls.update({ ...ctx.repo(), pull_number: pr.number, body: newBody });
-        updatePrPromise.catch((err) => log.error(err, "Error while updating PR body", { contextURL }));
+        updatePrPromise.catch((err) => log.error("Error while updating PR body", err, { contextURL }));
```
Comment (svenefftinge) [inline components/server/src/prebuilds/github-app.ts:573]: updating the type signatures of our log messages revealed that we were passing the arguments in the wrong order

File: components/gitpod-protocol/src/messaging/proxy-factory.ts (hunk 2 above)
Comment (akosyakov) [reply components/gitpod-protocol/src/messaging/proxy-factory.ts:232]: I think you can also do `resultPromise.then(resolve, reject)`

File: components/gitpod-protocol/src/util/gitpod-host-url.ts [hunk 4]
```diff
@@ -24,11 +24,10 @@ const workspaceUrlPrefixRegex = RegExp(`^(([0-9]{4,6}|debug)-)?${baseWorkspaceID
 export class GitpodHostUrl {
     readonly url: URL;
 
-    constructor(url: string) {
-        const urlParam = url as any;
+    constructor(urlParam: string | URL) {
```
Comment (akosyakov) [inline components/gitpod-protocol/src/util/gitpod-host-url.ts:27]: I don't remember why but I believe we removed union type here on purpose before. And the only client which can pass URL inside it is this class itself for `with` functions

File: components/server/src/prebuilds/github-app.ts (hunk 3 above)
Comment (akosyakov) [reply components/server/src/prebuilds/github-app.ts:573]: maybe we should enable linting rule to disable usage of console on server

Comment (svenefftinge) [reply components/server/src/prebuilds/github-app.ts:573]: yes, would make sense in a different PR

**Review by akosyakov — APPROVED** [review]

**Pull request merged**